[Spacetime] TaskEither Abstract Data Type for fallible async ops and …

wayneseymour opened the pull request:
…&& fromNullable pattern in TS

resolves: https://github.com/elastic/kibana-team/issues/1524

wayneseymour committed:
[Spacetime] TaskEither Abstract Data Type for fallible async ops and && fromNullable pattern in TS

resolves: https://github.com/elastic/kibana-team/issues/1524

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

wayneseymour committed:
huh?

wayneseymour committed:
fixed

wayneseymour committed:
going a different route, trying again with effect as I've learned that the author of fp-ts has joined the EffectTS team

wayneseymour committed:
Drop this

wayneseymour committed:
Add effect ts (aka fp-ts 3.0)

wayneseymour committed:
run effect ts experiments

wayneseymour committed:
Hrmm, well it kinda works but I think we can do better

wayneseymour committed:
now, it's not succeeding at all.

wayneseymour committed:
Working now.

wayneseymour committed:
lil cleanup

wayneseymour committed:
progress

wayneseymour committed:
so, it looks like getRule has to be a thunk

wayneseymour committed:
Feels like a lot more could be done but it's working.

wayneseymour committed:
Interesting, so runFork is a background thing.

wayneseymour committed:
Next...

wayneseymour committed:
The first type is the return type, the second is the error or errors that can be thrown.
The last type as 'never' means all the requireiments for the computation are satisfied.
e.g:

const tre: Effect.Effect<'tre', never, never> = Effect.succeed('tre' as const);
vs
const tre: Effect.Effect<string, never, never> = Effect.succeed('tre');

wayneseymour committed:
This commit is mostly about from when I was going from one dependent svc, to two.

wayneseymour committed:
so now we have provided impl's for both

wayneseymour committed:
cleaned it up a bit to make it easier to understand from the cli
